Mathlib/Data/Matroid/DirectSum.lean

```diff
@@ -0,0 +1,269 @@
+    (hI : indepDirectSum hME I) :
+    M₁.Indep (I ∩ M₁.E) := by
+  obtain ⟨I₁, I₂, rfl, hI₁, hI₂⟩ := hI
+  convert M₁.indep_subset hI₁ (Set.inter_subset_right M₁.E I₁) using 1
+  exact Set.union_inter_supset_left_of_disjoint hME (M₂.subset_ground I₂ hI₂)
+
+lemma indepDirectSum.rightIndep {M₁ M₂ : IndepMatroid α} {hME : M₁.E ∩ M₂.E = ∅} {I : Set α}
+    (hI : indepDirectSum hME I) :
+    M₂.Indep (I ∩ M₂.E) := by
+  obtain ⟨I₁, I₂, rfl, hI₁, hI₂⟩ := hI
+  convert M₂.indep_subset hI₂ (Set.inter_subset_right M₂.E I₂) using 1
+  exact Set.union_inter_supset_right_of_disjoint hME (M₁.subset_ground I₁ hI₁)
+
+lemma indepDirectSum_iff {M₁ M₂ : IndepMatroid α} (hME : M₁.E ∩ M₂.E = ∅)
+    {I : Set α} (hI : I ⊆ M₁.E ∪ M₂.E) :
+    indepDirectSum hME I ↔ M₁.Indep (I ∩ M₁.E) ∧ M₂.Indep (I ∩ M₂.E) :=
+  ⟨fun hid => ⟨hid.leftIndep, hid.rightIndep⟩, fun ⟨hM₁, hM₂⟩ => ⟨I ∩ M₁.E, I ∩ M₂.E, by aesop⟩⟩
+
+lemma maximals_indepDirectSum_iff {M₁ M₂ : IndepMatroid α}
+    (hME : M₁.E ∩ M₂.E = ∅) (I : Set α) :
+    I ∈ maximals (· ⊆ ·) {I | indepDirectSum hME I} ↔
+      I ⊆ M₁.E ∪ M₂.E ∧
+        I ∩ M₁.E ∈ maximals (· ⊆ ·) M₁.Indep ∧
+        I ∩ M₂.E ∈ maximals (· ⊆ ·) M₂.Indep := by
+  dsimp only [maximals, Set.mem_setOf_eq]
+  constructor
+  · intro hyp
+    have I_grounded : I ⊆ M₁.E ∪ M₂.E := hyp.left.ground
+    rw [indepDirectSum_iff hME I_grounded] at hyp
+    obtain ⟨⟨hM₁, hM₂⟩, hB⟩ := hyp
+    have I_as : I = I ∩ M₁.E ∪ I ∩ M₂.E := (Set.union_inters_of_subset_union I_grounded).symm
+    constructor
+    · exact I_grounded
+    constructor
+    · constructor
+      · exact hM₁
+      · intro B₁ hB₁ hI₁
+        rw [I_as, Set.union_inter_distrib_right, Set.inter_assoc, Set.inter_assoc,
+            Set.inter_comm M₂.E, hME, Set.inter_empty, Set.union_empty, Set.inter_self]
+        exact Set.subset_inter_of_redundant_right
+          (hB ⟨B₁, I ∩ M₂.E, rfl, hB₁, hM₂⟩ (I_as ▸ Set.union_inter_aux hI₁))
+          (M₁.subset_ground _ hB₁)
+    · constructor
+      · exact hM₂
+      · intro B₂ hB₂ hI₂
+        rw [I_as, Set.union_inter_distrib_right, Set.inter_assoc, Set.inter_assoc,
+            hME, Set.inter_empty, Set.empty_union, Set.inter_self]
+        exact Set.subset_inter_of_redundant_left
+          (hB ⟨I ∩ M₁.E, B₂, rfl, hM₁, hB₂⟩ (I_as ▸ Set.inter_union_aux hI₂))
+          (M₂.subset_ground _ hB₂)
+  · intro ⟨I_grounded, ⟨hI₁, hB₁⟩, ⟨hI₂, hB₂⟩⟩
+    have I_indep : indepDirectSum hME I :=
+      ⟨_, _, Set.union_inters_of_subset_union I_grounded, hI₁, hI₂⟩
+    rw [indepDirectSum_iff hME I_indep.ground]
+    constructor
+    · exact ⟨hI₁, hI₂⟩
+    · intro B hB hIB
+      rw [Set.subset_iff_subsets_of_disjoint hB.ground]
+      constructor
+      · exact hB₁ hB.leftIndep (Set.inter_subset_inter_left M₁.E hIB)
+      · exact hB₂ hB.rightIndep (Set.inter_subset_inter_left M₂.E hIB)
+
+/-- Direct sum of matroids as a matroid. -/
+def indepMatroidDirectSum {M₁ M₂ : IndepMatroid α} (hME : M₁.E ∩ M₂.E = ∅) : IndepMatroid α :=
```

Comment:
The intention with `IndepMatroid` was just that is it is an intermediate step towards defining a `Matroid` with alternative axioms. It makes sense to define direct sum using `IndepMatroid` internally, but the final input and output types should be `Matroid`. 

You can do this using `IndepMatroid.matroid`

Reply:
Thanks!

Reply:
All right, I added `matroidDirectSum`. Should `indepMatroidDirectSum` be changed to `private def` now?

Reply:
If so, I will remove `indepMatroidDirectSum` from the module docstring.

Reply:
it can probably be `Matroid.directSum`. And yes, `indepMatroidDirectSum` should either be made private or completely removed - you can just have the code that you used to definite it go inside `Matroid.directSum`.

Reply:
I made it private. Should other (even more low-level) definitions and lemmas be also private?